### ISSUE TITLE
Refactor - Consistent internal data structure

### ISSUE THREAD

**Issue by sytong** [buruzaemon/frequent-algorithm#24]
Currently the data structure of the Summary (S) will change from hash to a 2D Array after finding the K Most Frequent items (Step 2 of the algorithm). 

This is a quick way to find the K Most Frequent from a hash by sorting items by the keys but it leads to change of internal data structure.

I can revert that back to a hash actually but I am not sure if it will have performance impact to the algo.

One thing we can also do to investigate this is to set up some benchmarking (as a test).

Thoughts?

**Comment by buruzaemon** [comment]
I prefer accuracy and efficiency of the algorithm over speed. Let's not worry about benchmarking at this point.

**Comment by buruzaemon** [comment]
Let's shoot for doing this in time for end-of-month, as 0.0.3 release.

**Comment by buruzaemon** [comment]
After giving the Demaine et al paper another read, I have the following thoughts:
1. The summaries in the internal queue do not need to be sorted.
2. Hence, we do not need to know the index of the kth largest value.
3. If a function name is actually longer than the logic contained within, perhaps it is not needed?

**Comment by sytong** [comment]
> 1. The summaries in the internal queue do not need to be sorted.
> 2.  Hence, we do not need to know the index of the kth largest value.

I supposed the best way to describe it is: the kth-largest needs to be FOUND but sorting the whole list is not the best way to achieve this. If the quickselect method can do the same thing but has better performance once the list (size of **b**?) grows biggest then I will go for it.

> 1.   If a function name is actually longer than the logic contained within, perhaps it is not needed?

I am ok to do that actually.
The following is what I was thinking when I wrote the code:
In my mind, `@queue.length > @n/@b` means "I should take the first summary from the queue now as its size is bigger than n/b".
I wrote a method for that because **n** and **b** do not tell me what they are.

**Comment by sytong** [comment]
Will try to use quickselect to replace the brute force sorting and also the weird `[[][]]` data structure.

**Comment by sytong** [comment]
Problem:

```
raise ArgumentError.new(ERR_BADK) if k < 1 or k > list.size
```

This can't handle the case when the number of unique elements in a summary < k
e.g. for b = 10 and k = 3, [1 1 1 1 1 1 2 2 2 2]

We can still throw error if `k < 1` (obviously!!)
But if `k > list.size`, we can just return the last element of the list?

**Comment by sytong** [comment]
Submitted Pull Request https://github.com/buruzaemon/frequent-algorithm/pull/31 to fix this.

**Comment by buruzaemon** [comment]
I am actually on the fence on this one... choosing an approach that does not assume uniqueness is far easier to implement. For example, the present implementation will return [2,2,2] if k=3 for the case you gave above (see the unit tests). I believe this to be proper behavior, as ties for the top k slots are more than probable.

**Comment by buruzaemon** [comment]
Actually, I take that back.. the more I think about it, the more I think the present implementation is wrong. Let's think about enumerating the cases, and implement this correctly.

**Comment by sytong** [comment]
I am closing this as this is fixed/refactored as alongside Issue #33.
